server/quota/quota.go

```diff
@@ -0,0 +1,117 @@
+package quota
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	api "github.com/tigrisdata/tigris/api/server/v1"
+	"github.com/tigrisdata/tigris/server/config"
+	"github.com/tigrisdata/tigris/server/metadata"
+	"github.com/tigrisdata/tigris/server/transaction"
+	"go.uber.org/atomic"
+	"golang.org/x/time/rate"
+)
+
+var (
+	sizeLimitUpdateInterval int64 = 5 // seconds
+
+	ErrRateExceeded        = api.Errorf(api.Code_RESOURCE_EXHAUSTED, "request rate limit exceeded")
+	ErrThroughputExceeded  = api.Errorf(api.Code_RESOURCE_EXHAUSTED, "request throughput limit exceeded")
+	ErrStorageSizeExceeded = api.Errorf(api.Code_RESOURCE_EXHAUSTED, "data size limit exceeded")
+)
+
+type State struct {
+	Rate         *rate.Limiter
+	Throughput   *rate.Limiter
+	Size         atomic.Int64
+	SizeUpdateAt atomic.Int64
+	SizeLock     sync.Mutex
+}
+
+type Manager struct {
+	quota     sync.Map
+	cfg       *config.QuotaConfig
+	tenantMgr *metadata.TenantManager
+	txMgr     *transaction.Manager
+}
+
+var mgr Manager
+
+func Init(t *metadata.TenantManager, tx *transaction.Manager, c *config.QuotaConfig) {
+	mgr = *newManager(t, tx, c)
+}
+
+func Check(ctx context.Context, namespace string, size int) error {
```

Comment:
Check --> Allowed? Btw, who will be passing size? and what is this size?

Reply:
Will rename to "Allow".
Size is request payload size. We are limiting write throughput using it.

Reply:
should we call it payload then? or reqSize?